### ISSUE TITLE
[RRFC] <title>

### ISSUE THREAD

**Issue by osher** [npm/rfcs#549]
## Motivation ("The Why")
Consider a corporate with a codebase with a high coverage and a mature CI.
Consider that the corporate's policy is to test on latest and greatest and ship latest and greatest to enjoy all security patches published by the community.
Now consider a counter requirement not to use 3rd-party versions that are too young, because they have not spent enough time in the market for their exploits to be noted.

This requires `npm i` to distinct between packages of the organization, who's new versions should be accepted as soon as they are published, in contrast to 3rd party open-source packages, who's versions should be X days old before it can enter a build (given that no known issues are detected on them).

This requires a further evolution over the `npm i --before <date>` option
We need `npm` to be able to distinct between corporate proprietary packages and the rest.

This could be using scopes, however, I meet too many codebases that still use prefixes which are not scopes, so I believe regex or at least prefixes to be a better solution.

It will also be useful to configure the behavior for not finding a version old enough. Some cases it would be preferable to fail, others to warn.

### Example
Assuming we go with regex - then:

```
npm config set stabilityDays 7d
npm config set 'onOnlyTooNew' warn
npm config set '@my-corp/*.stabilityDays' 0
npm config set '@my-corp/*.onOnlyTooNew' info
npm config set package-that-gave-problems-before.stabilityDays 90
npm config set 'package-that-gave-problems-before.onOnlyTooNew' error
```

#### Current Behaviour
The current behavior does not distinct between packages.
There is a `--before <date>` switch, and it will reject too new versions - but it will know to let corporate packages to pass, inhibiting quick responses.

#### Desired Behaviour
Consider the example above - then:
 - by default - versions newer than 7 days are ignored.
 - packages who's name start with `@my-corp/` (actually match the pattern `/@my-corp\/.*/) can pass through
 - `package-that-gave-problems-before` only versions 90 days or older

### References
* https://github.com/npm/cli/issues/4494

**Comment by osher** [comment]
I tried to find such config for jfrog because that's what the team I'm consulting now uses, but did not find a way to do that. Same for gitlab package registry.

I don't know about nexus.

Here's a feature-request I started for verdaccio:
https://github.com/verdaccio/verdaccio/discussions/3029

**Comment by ljharb** [comment]
I’m more arguing that since the use case is uncommon - waiting a length of time doesn’t increase security, and if such a practice became common then it would just delay discovery of vulnerabilities until after that time anyways - it should be implemented in your private registry instead of in npm itself.

**Comment by osher** [comment]
I did not miss your argument, and I see where it comes from. However - `npm` already supports `--before`, and anything you said is true about `--before` as well, but yet, `--before/enjoy-by` is supported.

The feature request is not about if to wait or not - but to be able apply a different policy to proprietary packages and 3rd party.

**Comment by ljharb** [comment]
The before feature is massively useful to go “back in time” and debug when a problem started occurring; it’s got nothing to do with security concerns or waiting to update. I’ve used it extensively in the years it’s existed for this purpose alone.

**Comment by zkochan** [comment]
I agree that the `before` feature is not super useful from security perspective but it is very useful for caching package metadata. The current resolution mode demands the highest version that satisfies the range, so we need to always request a new metadata file and there's no way to use a metadata file from a local cache.

I have added a new resolution mode to pnpm called ["time-based"](https://pnpm.io/npmrc#resolution-mode). This resolution mode uses time to resolve versions and allows to leverage a local cache better. Related RFC: https://github.com/pnpm/rfcs/pull/2.

This resolution mode works perfectly. I have been using it for about two months already. The only issue is that the abbreviated metadata returned by the npm registry doesn't currently include the time field. So, when pnpm uses this resolution mode, it needs to request the non-abbreviated metadata. This makes installation slower with cold cache. It would be great if npm could return the time field with the abbreviated metadata.
